horovod/spark/lightning/datamodule.py

```diff
@@ -0,0 +1,112 @@
+                 cur_shard: int=0, shard_count: int=1, schema_fields=None, storage_options=None,
+                 steps_per_epoch_train: int=1, steps_per_epoch_val: int=1, verbose=True, **kwargs):
+        super().__init__()
+        self.train_dir = train_dir
+        self.val_dir = val_dir
+        self.num_train_epochs = num_train_epochs
+        self.has_val = has_val
+        self.train_batch_size = train_batch_size
+        self.val_batch_size = val_batch_size
+        self.shuffle_size = shuffle_size
+        self.num_reader_epochs = num_reader_epochs
+        self.reader_pool_type = reader_pool_type
+        self.reader_worker_count = reader_worker_count
+        self.transform_spec = transform_spec
+        self.inmemory_cache_all = inmemory_cache_all
+        self.cur_shard = cur_shard
+        self.shard_count = shard_count
+        self.schema_fields = schema_fields
+        self.storage_options = storage_options
+        self.steps_per_epoch_train = steps_per_epoch_train
+        self.steps_per_epoch_val = steps_per_epoch_val
+        self.verbose = verbose
+
+    def setup(self, stage=None):
+        # Assign train/val datasets for use in dataloaders
+        if stage == 'fit':
+            transform_spec = TransformSpec(self.transform_spec) if self.transform_spec else None
+            # In general, make_batch_reader is faster than make_reader for reading the dataset.
+            # However, we found out that make_reader performs data transformations much faster than
+            # make_batch_reader with parallel worker processes. Therefore, the default reader
+            # we choose is make_batch_reader unless there are data transformations.
+            if transform_spec:
+                reader_factory = make_reader
+            else:
+                reader_factory = make_batch_reader
+
+            self.train_reader = reader_factory(self.train_dir, num_epochs=self.num_reader_epochs,
+                                               cur_shard=self.cur_shard, shard_count=self.shard_count,
+                                               hdfs_driver=PETASTORM_HDFS_DRIVER,
+                                               schema_fields=self.schema_fields,
+                                               storage_options=self.storage_options)
+            if self.has_val:
+                self.val_reader = reader_factory(self.val_dir, num_epochs=self.num_reader_epochs,
+                                                 cur_shard=self.cur_shard, shard_count=self.shard_count,
+                                                 hdfs_driver=PETASTORM_HDFS_DRIVER,
+                                                 schema_fields=self.schema_fields,
+                                                 storage_options=self.storage_options)
+
+    def teardown(self, stage=None):
+        if stage == "fit" and self.verbose:
+            print("Tear down petastorm readers")
+            if not self.inmemory_cache_all:
```

Comment:
Why in memory cache do not need to do "stop" and "join"?

Reply:
Reader was used once, and was stopped in the last step of initializing inmemory dataloder.